From 1ef8d8840d9aad21e179facd0d1976ab9f05e654 Mon Sep 17 00:00:00 2001
From: Koen Kooi <koen@dominion.thruhere.net>
Date: Thu, 6 Jan 2011 13:23:45 +0100
Subject: [PATCH 1/2] OMAP3: beagle C4: enable upto 720MHz OPP

Beagle C4 uses a recent 3530 and the board design allows enabling 720MHz
OPP. tweak the default table to allow for higher OPP tables

Signed-off-by: Koen Kooi <koen@dominion.thruhere.net>
---
 arch/arm/mach-omap2/board-omap3beagle.c |   35 +++++++++++++++++++++++++++++++
 1 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/board-omap3beagle.c b/arch/arm/mach-omap2/board-omap3beagle.c
index 9393e75..a0462b9 100644
--- a/arch/arm/mach-omap2/board-omap3beagle.c
+++ b/arch/arm/mach-omap2/board-omap3beagle.c
@@ -829,6 +829,41 @@ static void __init beagle_opp_init(void)
 			pr_err("%s: turbo OPPs enabled!\n", __func__);
 		}
 	}
+	/* Custom OPP enabled for C4 */
+	if (omap3_beagle_get_rev() == OMAP3BEAGLE_BOARD_C4) {
+		struct omap_hwmod *mh = omap_hwmod_lookup("mpu");
+		struct omap_hwmod *dh = omap_hwmod_lookup("iva");
+		struct device *dev;
+
+		if (!mh || !dh) {
+			pr_err("%s: Aiee.. no mpu/dsp devices? %p %p\n",
+				__func__, mh, dh);
+			r = -EINVAL;
+		} else {
+			/* Enable MPU 720MHz */
+			dev = &mh->od->pdev.dev;
+			r = opp_enable(dev, 720000000);
+
+			/* Enable IVA 520MHz and lower opps */
+			dev = &dh->od->pdev.dev;
+			r |= opp_enable(dev, 520000000);
+		}
+		if (r) {
+			pr_err("%s: failed to enable higher opp %d\n",
+				__func__, r);
+			/*
+			 * Cleanup - disable the higher freqs - we dont care
+			 * about the results
+			 */
+			dev = &mh->od->pdev.dev;
+			opp_disable(dev, 720000000);
+			dev = &dh->od->pdev.dev;
+			opp_disable(dev, 520000000);
+		} else {
+			pr_err("%s: 720MHz MPU OPPs enabled!\n", __func__);
+		}
+	}
+	
 }
 
 static void __init omap3_beagle_init(void)
-- 
1.6.6.1

